tests/unittests/config/test_cc_ubuntu_advantage.py

```diff
@@ -197,160 +210,218 @@ def test_schema_validation(self, config, error_msg):
-    def test_handle_prefers_new_style_config(self, m_configure_ua):
-        """ubuntu_advantage should be preferred over ubuntu-advantage"""
-        cfg = {
-            "ubuntu-advantage": {"token": "nope", "enable": ["wrong"]},
-            "ubuntu_advantage": {"token": "token", "enable": ["esm"]},
-        }
-        handle("nomatter", cfg=cfg, cloud=None, log=self.logger, args=None)
-        self.assertEqual(
-            'WARNING: Deprecated configuration key "ubuntu-advantage"'
-            ' provided. Expected underscore delimited "ubuntu_advantage";'
-            " will attempt to continue.",
-            self.logs.getvalue().splitlines()[0],
-        )
-        m_configure_ua.assert_called_once_with(token="token", enable=["esm"])
-
-
-class TestMaybeInstallUATools(CiTestCase):
-
-    with_logs = True
+        with pytest.raises(RuntimeError, match=match):
+            handle("nomatter", cfg=cfg, log=mock.Mock(), **handle_kwargs)
 
-    def setUp(self):
-        super(TestMaybeInstallUATools, self).setUp()
-        self.tmp = self.tmp_dir()
 
-    @mock.patch("%s.subp.which" % MPATH)
-    def test_maybe_install_ua_tools_noop_when_ua_tools_present(self, m_which):
-        """Do nothing if ubuntu-advantage-tools already exists."""
-        m_which.return_value = "/usr/bin/ua"  # already installed
-        distro = mock.MagicMock()
-        distro.update_package_sources.side_effect = RuntimeError(
-            "Some apt error"
-        )
-        maybe_install_ua_tools(cloud=FakeCloud(distro))  # No RuntimeError
-
-    @mock.patch("%s.subp.which" % MPATH)
-    def test_maybe_install_ua_tools_raises_update_errors(self, m_which):
-        """maybe_install_ua_tools logs and raises apt update errors."""
-        m_which.return_value = None
-        distro = mock.MagicMock()
-        distro.update_package_sources.side_effect = RuntimeError(
-            "Some apt error"
-        )
-        with self.assertRaises(RuntimeError) as context_manager:
-            maybe_install_ua_tools(cloud=FakeCloud(distro))
-        self.assertEqual("Some apt error", str(context_manager.exception))
-        self.assertIn("Package update failed\nTraceback", self.logs.getvalue())
-
-    @mock.patch("%s.subp.which" % MPATH)
-    def test_maybe_install_ua_raises_install_errors(self, m_which):
-        """maybe_install_ua_tools logs and raises package install errors."""
-        m_which.return_value = None
+@mock.patch(f"{MPATH}.subp.which")
+class TestMaybeInstallUATools:
+    @pytest.mark.parametrize(
+        [
+            "which_return",
+            "update_side_effect",
+            "install_side_effect",
+            "expectation",
+            "log_msg",
+        ],
+        [
+            # Do nothing if ubuntu-advantage-tools already exists.
+            pytest.param(
+                "/usr/bin/ua",  # already installed
+                RuntimeError("Some apt error"),
+                None,
+                does_not_raise(),  # No RuntimeError
```

Comment:
BTW really like your `does_not_raise` context mgr. It really simplifies test validation.

Reply:
:) I took inspiration from [parametrizing-conditional-raising](https://docs.pytest.org/en/6.2.x/example/parametrize.html?highlight=does_not_raise#parametrizing-conditional-raising) and yes, it makes test cases easier to read. Thanks!